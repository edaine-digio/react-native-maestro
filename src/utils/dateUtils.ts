import dayjs from 'dayjs'

export const formatDate = (input: string) => {
  return dayjs(input).format('DD MMM YYYY')
}
