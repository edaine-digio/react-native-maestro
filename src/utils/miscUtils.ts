import { PostObj } from 'src/features/home/feed/Feed.component'

export const sortPosts = (posts: PostObj[]) =>
  posts.sort(
    (a: PostObj, b: PostObj) =>
      Date.parse(b.datePosted) - Date.parse(a.datePosted)
  )
