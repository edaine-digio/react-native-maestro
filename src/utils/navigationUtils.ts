import { NavigatorScreenParams } from '@react-navigation/native'

export type RootStackParamList = {
  Launch: undefined
  RegisterModal: undefined
  AppNav: undefined
  Splash: undefined
}

export enum RootStackRoutes {
  Launch = 'Launch',
  ReigsterModal = 'RegisterModal',
  AppNav = 'AppNav',
  Splash = 'Splash'
}

export type TabParamList = {
  Home: undefined
  Profile: undefined
  Settings: NavigatorScreenParams<RootStackParamList>
}

export enum TabRoutes {
  Home = 'Home',
  Profile = 'Profile',
  Settings = 'Settings'
}
