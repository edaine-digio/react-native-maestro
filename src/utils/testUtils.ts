export const mockNavigationProp: any = {
  navigate: jest.fn()
}

export const mockRouteProp: any = jest.fn()
