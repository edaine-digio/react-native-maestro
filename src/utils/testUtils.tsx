import { PreloadedState, configureStore } from '@reduxjs/toolkit'
import { RenderOptions, render } from '@testing-library/react-native'
import { PropsWithChildren } from 'react'
import { Provider } from 'react-redux'
import userReducer from 'src/store/slices/userSlice'
import { AppStore, RootState } from 'src/store/store'

// Since these "props" are being used across multiple screen tests, they are typed as any in
// the interest of simplicity, as each type is annotated on a per-screen basis and we only
// test the functions are called with a particular route name
export const mockNavigationProp: any = {
  navigate: jest.fn()
}

export const mockRouteProp: any = jest.fn()

// Wrapper for components and screens using global state for testing from Redux Toolkit docs
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = configureStore({ reducer: { user: userReducer }, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const Wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => {
    return <Provider store={store}>{children}</Provider>
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}
