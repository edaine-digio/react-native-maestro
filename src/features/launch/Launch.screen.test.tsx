import React from 'react'
import { fireEvent, screen, waitFor } from '@testing-library/react-native'
import {
  mockNavigationProp,
  mockRouteProp,
  renderWithProviders
} from 'src/utils/testUtils'
import { RootStackRoutes } from 'src/utils/navigationUtils'
import { Launch } from 'src/features/launch/Launch.screen'

describe('Launch', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  test('Should render as expected', () => {
    renderWithProviders(
      <Launch navigation={mockNavigationProp} route={mockRouteProp} />
    )

    expect(screen.toJSON()).toMatchSnapshot()
  })

  test('login() should update state with returned body, and subsequently navigate to the guts of the app', async () => {
    fetchMock.resetMocks()

    fetchMock.mockResponse(
      JSON.stringify({
        first_name: 'John',
        last_name: 'Citizen',
        email: 'john@example.com',
        token: 'token'
      })
    )

    const { store, getByTestId, getByText } = renderWithProviders(
      <Launch navigation={mockNavigationProp} route={mockRouteProp} />
    )

    fireEvent.changeText(getByTestId('loginInput'), 'john@example.com')
    fireEvent.changeText(getByTestId('passwordInput'), 'password')

    fireEvent.press(getByText('Login'))

    await waitFor(() => {
      const state = store.getState()

      expect(state.user).toEqual({
        firstName: 'John',
        lastName: 'Citizen',
        email: 'john@example.com',
        token: 'token',
        error: false,
        loading: false
      })
    })

    await waitFor(() => {
      expect(mockNavigationProp.navigate).toHaveBeenCalledWith(
        RootStackRoutes.RootNavigation
      )
    })
  })

  test('Login should stay disabled when username is not long enough', async () => {
    const { getByText, getByTestId } = renderWithProviders(
      <Launch navigation={mockNavigationProp} route={mockRouteProp} />
    )

    const loginInput = getByTestId('loginInput')

    fireEvent.changeText(loginInput, 'G')

    fireEvent.press(getByText('Login'))

    expect(mockNavigationProp.navigate).not.toHaveBeenCalledWith(
      RootStackRoutes.RootNavigation
    ) // handleLogin did not fire
  })
})
