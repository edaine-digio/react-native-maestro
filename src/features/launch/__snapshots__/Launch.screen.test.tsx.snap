// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Launch Should render as expected 1`] = `
<LinearGradient
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  colors={
    [
      "#FC9468",
      "#FFD1AD",
      "#F5EDE0",
    ]
  }
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        {
          "height": 80,
          "marginBottom": 32,
          "marginTop": 144,
          "width": "100%",
        }
      }
    >
      <DigioPrimary />
    </View>
    <View
      onLayout={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "marginBottom": 144,
            "paddingHorizontal": 48,
            "width": "100%",
          },
          {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View>
        <View
          style={
            {
              "alignItems": "flex-start",
            }
          }
        >
          <TextInput
            autoCapitalize="none"
            autoCorrect={false}
            onChangeText={[Function]}
            placeholder="username"
            placeholderTextColor="#8A8A8A"
            secureTextEntry={false}
            style={
              {
                "borderColor": "#243746",
                "borderRadius": 20,
                "borderWidth": 2,
                "color": "#243746",
                "fontFamily": "Archivo-Thin",
                "fontSize": 20,
                "marginBottom": 16,
                "padding": 12,
                "width": 200,
              }
            }
            testID="loginInput"
            value=""
          />
        </View>
        <View
          style={
            {
              "alignItems": "flex-start",
            }
          }
        >
          <TextInput
            autoCapitalize="none"
            autoCorrect={false}
            onChangeText={[Function]}
            placeholder="password"
            placeholderTextColor="#8A8A8A"
            secureTextEntry={true}
            style={
              {
                "borderColor": "#243746",
                "borderRadius": 20,
                "borderWidth": 2,
                "color": "#243746",
                "fontFamily": "Archivo-Thin",
                "fontSize": 20,
                "marginBottom": 16,
                "padding": 12,
                "width": 200,
              }
            }
            testID="passwordInput"
            value=""
          />
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignSelf": "center",
              "marginBottom": 16,
              "opacity": 1,
              "padding": 8,
            }
          }
        >
          <Text
            style={
              {
                "color": "#243746",
                "fontFamily": "Archivo-Thin",
                "fontSize": 32,
                "includeFontPadding": false,
                "textAlign": "center",
              }
            }
          >
            Login
          </Text>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignSelf": "center",
            "marginBottom": 16,
            "opacity": 1,
            "padding": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#243746",
              "fontFamily": "Archivo-Thin",
              "fontSize": 24,
              "includeFontPadding": false,
              "textAlign": "center",
            }
          }
        >
          Register
        </Text>
      </View>
    </View>
  </RCTSafeAreaView>
</LinearGradient>
`;
